cryptography/src/bls12381/primitives/ops.rs

```diff
@@ -499,32 +503,61 @@ where
         .collect::<Vec<_>>();
     let weights = compute_weights(indices)?;
 
-    // If concurrency is not required, recover signatures sequentially
-    let concurrency = std::cmp::min(concurrency, prepared_evals.len());
-    if concurrency == 1 {
-        return prepared_evals
-            .iter()
-            .map(|evals| {
-                threshold_signature_recover_with_weights::<V, _>(&weights, evals.iter().cloned())
-            })
-            .collect();
-    }
+    recover_signatures::<V>(weights, prepared_evals, concurrency)
+}
+
+#[allow(unused_variables)] // for no_std which ignores concurrency
+fn recover_signatures<'a, V>(
+    weights: BTreeMap<u32, Weight>,
+    prepared_evals: Vec<Vec<&Eval<V::Signature>>>,
+    concurrency: usize,
+) -> Result<Vec<V::Signature>, Error>
+where
+    V: Variant,
+    V::Signature: 'a,
+{
+    #[cfg(not(feature = "std"))]
+    return prepared_evals
+        .into_iter()
+        .map(|evals| {
+            threshold_signature_recover_with_weights::<V, _>(&weights, evals.iter().cloned())
+        })
+        .collect();
 
-    // Build a thread pool with the specified concurrency
-    let pool = ThreadPoolBuilder::new()
-        .num_threads(concurrency)
-        .build()
-        .expect("Unable to build thread pool");
-
-    // Recover signatures
-    pool.install(move || {
-        prepared_evals
-            .par_iter()
-            .map(|evals| {
-                threshold_signature_recover_with_weights::<V, _>(&weights, evals.iter().cloned())
-            })
-            .collect()
-    })
+    #[cfg(feature = "std")]
```

Comment:
There is an open issue to clean this all up #903, apologize it wasn't completed before you started.